src/lib/recurringPurposes.js

```diff
@@ -1,18 +1,248 @@
+import { CONTACTS_DOCTYPE } from 'src/doctypes'
+import {
+  getEndPlaceCoordinates,
+  getStartPlaceCoordinates,
+  setAutomaticPurpose
+} from 'src/lib/timeseries'
 import { getManualPurpose, getAutomaticPurpose } from 'src/lib/trips'
 import {
+  buildContactsWithGeoCoordinates,
   buildNewestRecurringTimeseriesQuery,
+  buildRecurringTimeseriesByStartAndEndPointRange,
   buildSettingsQuery,
   buildTimeseriesQueryByAccountIdAndDate,
-  builTimeserieQueryByAccountIdAndStartPlaceAndEndPlaceAndStartDateAndDistance,
+  queryContactByDocId,
   queryTimeserieByDocId
 } from 'src/queries/queries'
 
+import { models } from 'cozy-client'
 import log from 'cozy-logger'
 
+const { deltaLatitude, deltaLongitude, geodesicDistance } = models.geo
+
+const MAX_SPATIAL_THRESHOLD_M = 200
+
+const findContactsWithGeo = async client => {
+  const queryDef = buildContactsWithGeoCoordinates().definition
+  const contacts = await client.queryAll(queryDef)
+  return contacts
+}
+
+export const findClosestStartAndEnd = (timeserie, contacts) => {
+  const startCoordinates = getStartPlaceCoordinates(timeserie)
+  const endCoordinates = getEndPlaceCoordinates(timeserie)
+  let closestStart = {},
+    closestEnd = {}
+  let minStartDistance = Number.MAX_SAFE_INTEGER,
+    minEndDistance = Number.MAX_SAFE_INTEGER
+  for (const contact of contacts) {
+    for (const address of contact.address) {
+      const addressCoordinates = {
+        lon: address.geo.geo[0],
+        lat: address.geo.geo[1]
+      }
+      const startPlaceDistance = geodesicDistance(
+        startCoordinates,
+        addressCoordinates
+      )
+
+      const endPlaceDistance = geodesicDistance(
+        endCoordinates,
+        addressCoordinates
+      )
+
+      if (startPlaceDistance < minStartDistance) {
+        closestStart = {
+          distance: startPlaceDistance,
+          contact,
+          address,
+          newCoordinates: startCoordinates
+        }
+        minStartDistance = startPlaceDistance
+      }
+      if (endPlaceDistance < minEndDistance) {
+        closestEnd = {
+          distance: endPlaceDistance,
+          contact,
+          address,
+          newCoordinates: endCoordinates
+        }
+        minEndDistance = endPlaceDistance
+      }
+    }
+  }
+  return { closestStart, closestEnd }
+}
+
+const setAddressContactRelationShip = ({
+  timeserie,
+  contact,
+  addressId,
+  relType
+}) => {
+  const newRel = {
+    [relType]: {
+      data: {
+        _id: contact._id,
+        _type: CONTACTS_DOCTYPE,
+        metadata: {
+          addressId
+        }
+      }
+    }
+  }
+  const timeserieWithRel = {
+    ...timeserie,
+    relationships: { ...(timeserie.relationships || {}), ...newRel }
+  }
+
+  return timeserieWithRel
+}
+
+const updateAddressCoordinates = async ({
+  client,
+  newCoordinates,
+  contact,
+  addressId
+}) => {
+  // Query contact to get latest rev
+  const newContact = await queryContactByDocId(client, contact._id)
```

Comment:
Why do you do that instead of a try / catch if there is a 409?

Reply:
Because I did this quickly and naively, which is not good. Thanks!

Reply:
See [5c9816f](https://github.com/cozy/coachCO2/pull/293/commits/5c9816f6bd79b4fb1fb20a43995540436a9f639b)